---
title: "Genomic Peaks Analysis & Comparative Distribution"
subtitle: "Gene regulatory dynamics during embryogenesis in Cloeon dipterum"
author: "Maria Rossello"
date: "Last run: `r format(Sys.time(), '%d/%m/%Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: flatly
    highlight: tango
---

```{r opts_knit setup, include=FALSE, warning=FALSE}
# Get your working directory (where you have your data):
wd <- "/home/mariarossello/atac_paper_analysis/"
setwd(file.path(wd))

knitr::opts_knit$set(
  root.dir = normalizePath(wd))
```

# Setting up project folders and libraries

This section outlines the initial setup. It involves loading specific R functions, setting up necessary libraries, and organizing data directories to streamline file management and data processing.

*Loading necessary functions.* This code chunk is responsible for importing custom R functions necessary for processing and analyzing ATAC-seq data. These functions are stored externally and are essential for the specific analyses that will be conducted in subsequent steps. 

```{r message=FALSE, warning=FALSE}
# Load functions from a script located in the working directory for downstream analysis
source(paste0(wd,"/atac_downstrem_analysis_functions.R"))
```

*Loading required libraries.* A suite of R packages is loaded here, enabling data manipulation, visualization, and advanced statistical analysis

```{r message=FALSE, warning=FALSE}
# Load libraries for data manipulation
library(dplyr)        # Data manipulation functions such as filtering and arranging
library(tidyr)        # Data tidying and reshaping tools
library(tibble)       # Modern and flexible data frame object for data manipulation

# Load libraries for data input and string operations
library(readr)        # Reading and writing tabular data quickly and easily
library(stringr)      # String operations for data cleaning and preparation

# Load libraries for statistical analysis
library(edgeR)        # Analysis of differential expression of RNA-seq count data
library(purrr)        # Functional programming tools for working with lists and vectors

# Load libraries for data visualization
library(ggplot2)      # Data visualization using a grammar of graphics
library(scales)       # Customization of plot scales in ggplot2 for aesthetic tweaks
```

The input data directories are systematically organized to ensure easy access and management of data files necessary for the analysis:

- 0_EXPERIMENT_INFO: Contains metadata about the experiments, such as sample details, experimental conditions...
- 1_COUNTS: Stores count matrices.
- 2_PEAK_INFO: Includes information about genomic regions of open chromatin identified in the ATAC-seq data.
- 3_OTHER_SP_DATA: Holds comparative ATACseq data from other species that will be used for cross-species analysis. This data comes from publicly available data.
- 4_NATCOM2020_INFO: Specific mayfly genomic information coming from [Almudi _et al.,_ 2020](https://doi.org/10.1038/s41467-020-16284-8).
- 5_RNA-SEQ: Directory for storing RNA-seq data files.

We save the paths to this folders in variables to acces them during all the analysis

```{r folder with input data}
info  <- paste0(wd,"0_EXPERIMENT_INFO/")
counts  <- paste0(wd,"1_COUNTS/")
peak_info <- paste0(wd,"2_PEAK_INFO/")
sp_data <- paste0(wd,"3_OTHER_SP_DATA/")
natcom2020 <- paste0(wd,"4_NATCOM2020_INFO/")
rnaseq <- paste0(wd,"5_RNA-SEQ/")
```

The output data structure is geared towards organizing the results of the analysis into logical and functional directories, making it easier to locate and use the data in subsequent steps or for final reporting.

- A1_PEAK_ABUNDANCE: This is the main folder designated for storing all outputs related to differential peak analysis. It acts as a central hub for aggregating analytical results. This will be designed as our analysis folder.
- output_images: A subfolder created within the main analysis folder to store all graphical representations produced during the analysis. 
- output_files: This directory is reserved for statistical outputs, such as tables.

```{r folders to save results}
analysis_folder <- paste0(wd,"A1_PEAK_ABUNDANCE/")
imgdir  <- paste0(analysis_folder,"output_images/")
statdir <- paste0(analysis_folder,"output_files/")
```

```{r opts_chunk setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE,
	cache = TRUE,
	fig.path = imgdir,
	fig.width = 10,
	fig.height = 7
)
setwd(file.path(wd))
```

# Importing and organizing ATAC-seq data

## C. dipterum ATAC data

We'll start by loading the sample information file for the ATAC-seq analysis, which includes details for each sample.

```{r warning=FALSE}
samples_info <- data.frame(read_delim(paste0(info,"info_samples.csv"),
                           delim = ",", 
                           escape_double = FALSE, 
                           trim_ws = TRUE))
samples_info
```

Next, we consolidate the consensus peak counts, which are initially spread across different files for each developmental stage, into a single file for ease of analysis.

```{r}
# Specify the folder path
counts_path <- paste0(counts,"counts_stage/")
file_paths <- list.files(path = counts_path, full.names = TRUE, pattern = "\\.counts$")

# Use a costume function to merge the files
counts_df <- merge_counts_files(file_paths)

# Save the df for future analysis
write_delim(counts_df, paste0(counts,"sample_raw_counts.tsv"), "\t")

head(counts_df)
```

We also import peak information associated with genes and unique zones.

```{r}
cdip_peak_zone_associated_gene <- data.frame(read_delim(paste0(peak_info,"clodip_embryoATAC_peaks_and_genes.tsv"),
                                                        delim = "\t", escape_double = FALSE,
                                                        trim_ws = TRUE))
cdip_peak_zone_associated_gene <- cdip_peak_zone_associated_gene %>% rename(peak_id = peak)
rownames(cdip_peak_zone_associated_gene) <- cdip_peak_zone_associated_gene$peak_id
head(cdip_peak_zone_associated_gene)
```

```{r}
cdip_peak_unique_zone<- data.frame(read_delim(paste0(peak_info,"clodip_embryoATAC_peaks_and_zones.tsv"),
                                              delim = "\t", escape_double = FALSE,
                                              trim_ws = TRUE))

 cdip_peak_unique_zone <- cdip_peak_unique_zone %>% rename(peak_id = peak)
 rownames(cdip_peak_unique_zone) <- cdip_peak_unique_zone$peak_id
 head(cdip_peak_unique_zone)
```

We join the two previous tables:

```{r}
cdip_peak_associated_gene <- left_join(cdip_peak_unique_zone,
                                       cdip_peak_zone_associated_gene,
                                       by = "peak_id") %>%
  rowwise() %>%
  mutate(
    clodip_v3 = case_when(
      zone == "promoter" ~ promoter,
      zone == "proximal" ~ proximal,
      zone == "genebody" ~ gene_body,
      # Add more conditions here if there are other zones
      TRUE ~ NA_character_
    )
  ) %>%
  select(peak_id, clodip_v3, zone)

head(data.frame(cdip_peak_associated_gene))
```

```{r}
write_delim(counts_df, paste0(peak_info,"clodip_embryoATAC_peaks_associated_gene.tsv"), "\t")
```

## Other anaimals ATAC data

Similarly, we import peak information for *D. melanogaster*. We use the data provided in [Bozek et al,. 2019](https://doi.org/10.1101/gr.242362.118). 

```{r}
dmel_peak_unique_zone <- data.frame(read_delim(paste0(sp_data,"/dmel_embryoATAC_peaks_and_zones.tsv"),
                                               delim = "\t", escape_double = FALSE,
                                               trim_ws = TRUE))
dmel_peak_unique_zone <- dmel_peak_unique_zone %>% rename(peak_id = peak)
rownames(dmel_peak_unique_zone) <- dmel_peak_unique_zone$peak_id
head(dmel_peak_unique_zone)
```

From [MarlÃ©taz et al,. 2018](https://doi.org/10.1038/s41586-018-0734-6) we optin peak information for  *B. lanceolatum* and *D. rerio*

```{r}
# Specifying counts for other species
blan_zone <- c(distal = 30020, genebody = 21546, promoter = 23355, proximal = 13470)
drer_zone <- c(distal = 143097, genebody = 21020, promoter = 28798, proximal = 5582)
```

# Parameters Setup
This section allows you to define key variables for your experiment and set aesthetic parameters for your graphs, making your analysis both rigorous and visually coherent.

## Cut Off

We start by setting the thresholds for the analysis:

```{r}
# Setting minimum filter threshold
MIN_FILT = 10

cat(paste0("We consider less than ", MIN_FILT, " counts as no expression."))
```

## Conditions to Study

Here, we define the experimental conditions, which are crucial for differential analysis.

```{r}
# Creating a factor variable for the stages of development, which will be used in differential analysis
condition_factors <- factor(samples_info$stage,
                            levels=c("emb4","emb6","emb8","emb10", "emb12", "emb14"))

# Assigning the factor to a new column in samples_info for easy reference
samples_info$condition <- condition_factors

```

## Aesthetic Parameters

Important: To ensure your analysis runs smoothly and your graphs look exactly as intended, it's crucial that the names you use in the following lists exactly match those defined in your data files. This consistency is key for linking your data correctly with the aesthetic elements like colors.

Now, let's define some color parameters to keep our graphs looking consistent and visually appealing:

```{r fig.width=3, fig.height=3}
# Defining color schemes for the conditions, ensuring they match those in samples_info$condition
condition_colors <- c("emb4"="#0072B2",  
                      "emb6"="#009E73",  
                      "emb8"="#D55E00",  
                      "emb10"="#CC79A7", 
                      "emb12"="#F0E442", 
                      "emb14"="#56B4E9")

# Setting colors for individual samples within each condition
sample_colors <- c("e4_1"="#0072B2", 
                   "e4_2"="#72C2FF",  
                   "e6_1"="#009E73",
                   "e6_2"="#72FFB4",  
                   "e8_1"="#D55E00",
                   "e8_2"="#FFA64D", 
                   "e10_1"="#CC79A7",
                   "e10_2"="#FFB3D9", 
                   "e12_1"="#F0E442",
                   "e12_2"="#FFFF99",
                   "e14_1"="#56B4E9",
                   "e14_2"="#99D6FF")

# Color coding for different peak zones
zone_colors <- c("promoter"="#DDC50F", 
                 "proximal"="#228B22",  
                 "genebody"="#4169E1",
                 "distal"="#984EA3")

# Additional color settings for stages
stage_colors <- c("open"="#48A9A6",
                  "closed"="#C97C5D")

```

We also specify the order in which samples and stages should be arranged for visual consistency across figures:

```{r}
# Ordering of samples for plots
sample_order <- c("e4_1", 
                  "e4_2", 
                  "e6_1", 
                  "e6_2", 
                  "e8_1", 
                  "e8_2", 
                  "e10_1", 
                  "e10_2", 
                  "e12_1", 
                  "e12_2", 
                  "e14_1", 
                  "e14_2")

# Mapping of stages to descriptive names
stage_order <- c("emb4"="Stage 4",
                 "emb6"="Stage 6", 
                 "emb8"="Stage 8", 
                 "emb10"="Stage 10", 
                 "emb12"="Stage 12", 
                 "emb14"="Stage 14")
```

# Counts exploratory analysis

This section focuses on structuring and analyzing sequencing count data to facilitate a clear understanding of sample variability before and after normalization. The initial step involves encapsulating the count data and sample metadata into a DGEList object. This approach ensures organized data handling and simplifies subsequent analytical procedures.

```{r}
# Create a DGEList object with count data and sample information
dge <- DGEList(counts=as.matrix(filter_peakid(counts_df)),
               samples=samples_info,
               group=samples_info$condition)
```

Following the data structuring, library sizes are calculated and visualized to assess the initial distribution across samples.

```{r}
# Calculate and order library sizes for initial visualization
library_sizes_df <- data.frame(sample=colnames(dge$counts),
                            libsize=colSums(dge$counts))
library_sizes_df <- names_to_graph(library_sizes_df, "sample", sample_order)
```

```{r library_size_raw}
# Plotting the library sizes before normalization
libsize_plot <- ggplot(library_sizes_df, aes(x = sample, y = libsize)) +
  geom_bar(stat = "identity",  fill = sample_colors) +
  theme_classic() +
  labs(title = "Library Sizes Before Normalization",
       x = "Sample",
       y = "Library size")
print(libsize_plot)
```

```{r message=FALSE, warning=FALSE}
# Save the initial library size plot
save_ggplot(libsize_plot, "library_size_raw")
```

Discrepancies in library size prompt the application of normalization, utilizing the Trimmed Mean of M-values (TMM) method. TMM effectively adjusts for library size and compositional differences, enabling robust comparisons across samples.

```{r}
# Normalize library sizes using TMM
dge <- calcNormFactors(dge, method="TMM")

# Calculate and adjust the library sizes based on normalization factors
norm_factors <- dge$samples$norm.factors
norm_lib_sizes <- colSums(dge$counts) / norm_factors * mean(norm_factors)

# Generate normalized count data
norm_counts <- sweep(dge$counts, 2, norm_factors, "/") * mean(norm_factors)
norm_counts_df <- data.frame(norm_counts) %>% 
  tibble::rownames_to_column("peak_id") 

# Output normalized counts to a file
write_delim(norm_counts_df,
            paste0(counts,"sample_nomalized_counts.tsv"), 
            "\t")
```

Post-normalization, library sizes are re-evaluated and plotted to confirm the effectiveness of the normalization process.

```{r}
# Recalculate library sizes from normalized counts
library_sizes_norm_df <- data.frame(sample=colnames(norm_counts),
                            libsize=colSums(norm_counts))
library_sizes_norm_df <- names_to_graph(library_sizes_norm_df, "sample", sample_order)
```

```{r library_size_norm}
# Plotting the library sizes after normalization
libsize_norm_plot <- ggplot(library_sizes_norm_df, aes(x = sample, y = libsize)) +
  geom_bar(stat = "identity",  fill = sample_colors) +
  geom_point(data = library_sizes_df, aes(x = sample, y = libsize), color = "black", size = 3) +
  theme_classic() +
  labs(title = "Library Sizes After Normalization",
       x = "Sample",
       y = "Library size")
print(libsize_norm_plot)
```

```{r message=FALSE, warning=FALSE}
# Save the normalized library size plot
save_ggplot(libsize_norm_plot, "library_size_norm")
```

Lastly, the document outlines the aggregation of normalized counts by biological stages, facilitating a focused analysis on stage-specific trends and variations.

```{r}
# Aggregate normalized counts by sample stage
norm_counts_x_stage_stats_df <- norm_counts_df %>%
  pivot_longer(-peak_id, names_to = "sample", values_to = "norm_counts")  %>%
  mean_collapse_data_by_stage (samples_info=samples_info, df_to_collapse=., numeric_col="norm_counts")

# Extract and save the mean and standard deviation of normalized counts by stage
norm_counts_x_stage_mean_df <- norm_counts_x_stage_stats_df %>%
  pivot_wider(-sd, names_from = stage, values_from=mean) %>%
  data.frame()

# Get the mean normalized counts
norm_counts_x_stage_sd_df <- norm_counts_x_stage_stats_df %>%
  pivot_wider(-mean, names_from = stage, values_from=sd) %>%
  data.frame()

# Save the mean and SD data of normalized counts for each stage
write_delim(norm_counts_x_stage_mean_df,
            paste0(counts,"stage_normalized_counts.tsv"),
            "\t")
```

We see the library size after collapsing by embryonic stage.

```{r}
# Recalculate library sizes from normalized counts
library_sizes_stage_norm_df <- data.frame(stage=colnames(filter_peakid(norm_counts_x_stage_mean_df)),
                            libsize=colSums(filter_peakid(norm_counts_x_stage_mean_df)))
library_sizes_stage_norm_df <- names_to_graph(library_sizes_stage_norm_df, "stage", stage_order)
```

```{r library_size_stage}
# Plotting the library sizes after normalization
libsize_stage_plot <- ggplot(library_sizes_stage_norm_df, aes(x = stage, y = libsize)) +
  geom_bar(stat = "identity",  fill = condition_colors) +
  theme_classic() +
  labs(title = "Library Sizes After Normalization",
       x = "Sample",
       y = "Library size")
print(libsize_stage_plot)
```

```{r message=FALSE, warning=FALSE}
# Save the normalized library size plot
save_ggplot(libsize_stage_plot, "library_size_stage")
```

# Number of peaks analysis

In this analysis, we determine the presence of peaks based on count thresholds. Peaks with counts above a certain threshold (`r MIN_FILT`) are considered "open" (active), while those below are "closed" (inactive). We then analyze the global number of peaks and their dynamics throughout different conditions and stages.

```{r}
# Assessing whether each peak is open based on the MIN_FILT threshold
peak_is_open <- filter_peakid(norm_counts_df) > MIN_FILT
```

## Global number of peaks

We first look at the number of peaks present in each sample.

```{r}
# Calculating the number of open peaks per sample
n_x_sample_df <- stack(colSums(peak_is_open))
colnames(n_x_sample_df) <- c("n_peaks", "sample")
```

Next, we calculate the number of peaks per stage, considering the average across samples in each stage.

```{r}
# Aggregating data by stage and calculating means
n_x_stage_df <- mean_collapse_data_by_stage(samples_info, n_x_sample_df, numeric_col="n_peaks")
n_x_stage_df <- names_to_graph(n_x_stage_df, "stage", stage_order)
```

Visualizing the number of peaks per stage, including variability.

```{r num_peaks}
# Plotting the mean number of peaks per stage with standard deviation
n_x_stage_df_plot <- ggplot(n_x_stage_df, aes(x = stage, y = mean, group = 1)) +
  geom_ribbon(aes(ymin = mean - sd, ymax = mean + sd), fill = "grey60", alpha = 0.2) +
  geom_line(color = "black") +
  theme_classic() +
  labs(title = "Total number of peaks per stage", x = "Developmental stage", y = "Number of Peaks")
print(n_x_stage_df_plot)
```

```{r message=FALSE, warning=FALSE}
save_ggplot(n_x_stage_df_plot, "num_peaks")
```

## Peaks dynamics

### Stablising peaks dynamics

Now, we explore the dynamics of peaks, distinguishing between those that are always open, always closed, (consensus) or show variability (dynamic).

```{r}
# Assigning row names for easier identification
rownames(peak_is_open) <- norm_counts_df$peak_id

# Summarizing open and closed peaks
open_peaks <- colSums(peak_is_open)
closed_peaks <- nrow(peak_is_open) - open_peaks

# Preparing data for plotting
peak_x_sample <- data.frame(open = open_peaks, closed = closed_peaks)
peak_x_sample$sample <- rownames(peak_x_sample)
```

Identifying constitutive peaks that remain consistently open or closed across all stages

```{r}
# Identifying always open and always closed peaks
# Always TRUE in the peak_is_open matrix
always_open <-rowSums(peak_is_open) == ncol(peak_is_open)
# Always FLASE in the peak_is_open matrix
always_closed <-rowSums(peak_is_open) == 0

# Extracting names of constitutive peaks
constitutive_peaks_names <- c(names(always_open[always_open]), names(always_closed[always_closed]))
```

For this analysis we have `r length(constitutive_peaks_names)` constitutive where `r length(always_open[always_open])` are always open and `r length(always_closed[always_closed])` always closed.

That leaves us with `r nrow(peak_is_open)-length(constitutive_peaks_names)` dynamic peaks. 

We then focus on dynamic peaks, those that vary between stages or samples.

```{r}
# Filtering for dynamic peaks
dynamic_norm_counts_df <- norm_counts_df[!(norm_counts_df$peak_id %in% constitutive_peaks_names),]
dynamic_norm_counts_x_stage_df <- norm_counts_x_stage_mean_df[!(norm_counts_x_stage_mean_df$peak_id %in% constitutive_peaks_names),]
```

We save the constitutive and dynamic peak.
```{r}
# Raw counts
write_delim(counts_df[(norm_counts_df$peak_id %in% constitutive_peaks_names),], 
            paste0(counts,"constitutive_peaks_sample_raw.tsv"), "\t")
write_delim(counts_df[!(norm_counts_df$peak_id %in% constitutive_peaks_names),], paste0(counts,"dynamic_peaks_sample_raw.tsv"), "\t")

#Normalized counts x sample
write_delim(norm_counts_df[(norm_counts_df$peak_id %in% constitutive_peaks_names),], 
            paste0(counts,"constitutive_peaks_sample_norm.tsv"), "\t")
write_delim(norm_counts_df[!(norm_counts_df$peak_id %in% constitutive_peaks_names),], 
            paste0(counts,"dynamic_peaks_sample_norm.tsv"), "\t")

#Normalized counts x stage
write_delim(norm_counts_x_stage_mean_df[(norm_counts_x_stage_mean_df$peak_id %in% constitutive_peaks_names),], 
            paste0(counts,"/constitutive_peaks_stage.tsv"), "\t")
write_delim(norm_counts_x_stage_mean_df[!(norm_counts_x_stage_mean_df$peak_id %in% constitutive_peaks_names),], 
            paste0(counts,"/dynamic_peaks_stage.tsv"), "\t")
```

Assessing the dynamics of these peaks across samples.

```{r}
# Calculating open and closed peaks in dynamic sets
dynamics_x_sample_df <- data.frame(sample = colnames(filter_peakid(dynamic_norm_counts_df)),
                                   open_peaks = stack(colSums(filter_peakid(dynamic_norm_counts_df) > MIN_FILT))["values"],
                                   closed_peaks = stack(colSums(filter_peakid(dynamic_norm_counts_df) <= MIN_FILT))["values"])
colnames(dynamics_x_sample_df) = c("sample", "open", "closed")

# Pivoting data for plotting
dynamics_x_sample_df <- pivot_longer(dynamics_x_sample_df, cols = c(open, closed),
                                     names_to = "peak_status", values_to = "peaks")

```

Finally, we analyze and visualize these dynamics across stages.

```{r}
# Collapsing data by stage for dynamic peaks
dynamics_x_stage_dftoplot <- mean_collapse_data_by_stage(samples_info, dynamics_x_sample_df, numeric_col="peaks")
dynamics_x_stage_dftoplot <- names_to_graph(dynamics_x_stage_dftoplot, "stage", stage_order)
```

Visualizing peak dynamics per stage:

```{r num_dynamic_peaks}
# Plotting peak dynamics across stages
dynamics_x_stage_plot <- ggplot(dynamics_x_stage_dftoplot, aes(x = stage, y = mean, group = peak_status, color = peak_status)) +
  geom_ribbon(aes(ymin = mean - sd, ymax = mean + sd, fill = peak_status), alpha = 0.2, color = NA) +
  geom_line() +
  theme_classic() +
  labs(title = "Dynamics of peaks per stage",
       x = "Developmental stage", y = "Number of Peaks") +
  scale_fill_manual(values = stage_colors) +
  scale_color_manual(values = stage_colors)
print(dynamics_x_stage_plot)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
save_ggplot(dynamics_x_stage_plot, "num_dynamic_peaks")
```

# Peaks Per Zone Analysis

This section explores the distribution of open peaks across different genomic zones (e.g., promoter, gene body) for each stage of development, focusing first on all peaks and then specifically on dynamic peaks. Additionally, we compare these distributions across different species.

## Number of Zones per Stage

We start by determining the number of peaks per zone for each condition, working by stage rather than individual samples. This requires combining counts data (to check if a peak is expressed in a condition) with zone data.

```{r}
# Determine which peaks are open based on the MIN_FILT threshold
peak_is_open_x_stage <- filter_peakid(norm_counts_x_stage_mean_df) > MIN_FILT
```

Now, we aggregate the number of peaks per zone for each stage:

```{r}
# Count the number of peaks in each zone for each stage
zone_x_stage_df <- count_peaks_x_zones(peak_is_open_matrix=peak_is_open_x_stage, peak_zones_df=cdip_peak_unique_zone)
zone_x_stage_df <- names_to_graph(zone_x_stage_df, "stage", stage_order)
```

Visualizing the distribution of open peaks across zones for each stage:

```{r zones_stages}
# Plotting the proportion of peaks per zone for each stage
zone_x_stage_plot <- ggplot(zone_x_stage_df, aes(x = stage, 
                            y = number_of_peaks, 
                            fill = factor(zone, levels = c("distal", "genebody", "proximal", "promoter")), 
                            label = number_of_peaks)) +
  geom_bar(stat = "identity", alpha=0.8) +
  geom_text(position = position_stack(vjust = 0.5), size = 3) +
  scale_fill_manual(values = zone_colors) +
  scale_y_continuous() +
  theme_classic() +
  labs(y = "Percentage of Peaks", x = "Stage", fill = "Zone") +
  ggtitle("Peaks per zone")
print(zone_x_stage_plot)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
save_ggplot(zone_x_stage_plot, "zones_stages")
```


##Analyzing Dynamic Peaks

Next, we focus on dynamic peaks, identifying their distribution across zones for each stage:

```{r}
# Determine which dynamic peaks are open
dynamic_peak_is_open_x_stage <- filter_peakid(dynamic_norm_counts_x_stage_df) > MIN_FILT

# Count the number of dynamic peaks in each zone for each stage
dynamic_zone_x_stage_df <- count_peaks_x_zones(peak_is_open_matrix=dynamic_peak_is_open_x_stage, peak_zones_df=cdip_peak_unique_zone)
dynamic_zone_x_stage_df <- names_to_graph(dynamic_zone_x_stage_df, "stage", stage_order)
```

Visualizing dynamic peaks across zones per stage:

```{r zones_dynamic_peaks}
# Plotting the distribution of dynamic peaks across zones for each stage
ggplot(dynamic_zone_x_stage_df, aes(x = stage, 
                                     y = number_of_peaks, 
                                     fill = factor(zone, levels = c("distal", "genebody", "proximal", "promoter")), 
                                     label = number_of_peaks)) +
  geom_bar(stat = "identity", alpha=0.8) +
  geom_text(position = position_stack(vjust = 0.5), size = 3) +
  scale_fill_manual(values = zone_colors) +
  scale_y_continuous() +
  theme_classic() +
  labs(y = "Percentage of Peaks", x = "Stage", fill = "Zone") +
  ggtitle("Dynamic peaks per zone")
```

```{r message=FALSE, warning=FALSE, include=FALSE}
save_ggplot(dynamics_x_stage_plot, "zones_dynamic_peaks")
```

## Number of Zones Across Species

Finally, we compare the distribution of peaks across genomic zones between different species, including both constitutive and dynamic peaks:

```{r}
# Defining zone counts for C. dipterum
zones <- c("distal", "genebody", "promoter", "proximal")
cdip_constitutive_zone_table <- as.vector(table(cdip_peak_unique_zone[(cdip_peak_unique_zone$peak_id %in% constitutive_peaks_names),]$zone))
cdip_constitutive_zone <- as.vector(cdip_constitutive_zone_table)
names(cdip_constitutive_zone) <- names(cdip_constitutive_zone_table)

cdip_dynamic_zone_table <- as.vector(table(cdip_peak_unique_zone[!(cdip_peak_unique_zone$peak_id %in% constitutive_peaks_names),]$zone))
cdip_dynamic_zone <- as.vector(cdip_dynamic_zone_table)
names(cdip_dynamic_zone) <- names(cdip_dynamic_zone_table)

# Defining zone counts for D.melanogaster
dmel_zone_table <- table(dmel_peak_unique_zone$zone)
dmel_zone <- as.vector(dmel_zone_table)
names(dmel_zone) <- names(dmel_zone_table)
```

Preparing the data for visualization:

```{r}
# Preparing the dataset for comparison across species
zone_x_sp_df <- data.frame(cdip_constitutive=cdip_constitutive_zone,
                           cdip_dynamic=cdip_dynamic_zone,
                           dmel=dmel_zone,
                           blan=blan_zone,
                           drer=drer_zone,
                           zone=zones) %>%
  pivot_longer(cols = colnames(select(., -zone)),
               names_to = "type",
               values_to = "number_of_peaks") %>%
  group_by(type) %>%
  mutate(total_peaks = sum(number_of_peaks),
         proportion = number_of_peaks / total_peaks) %>%
  ungroup() %>%
  select(., -total_peaks)

```

Visualizing the comparative distribution of peaks across genomic zones among different species:

```{r zones_sp_comparative}
# Define the custom order of x-axis
sp_order <- c("cdip_constitutive"="C. dipterum\nConsitutive peaks",
                  "cdip_dynamic"="C. dipterum\nDynamic peaks", 
                  "dmel"="D. melanogaster", 
                  "blan"="B. lanceolatum",
                  "drer"="D. rerio")  

zone_x_sp_df <- names_to_graph(zone_x_sp_df, "type", sp_order)



# Plotting the proportion of peaks per zone across species
zone_x_sp_plot <- ggplot(zone_x_sp_df, aes(x = type,
                                           y = proportion, 
                                           fill = factor(zone, levels = c("distal", "genebody", "proximal", "promoter")), 
                                           label = number_of_peaks)) +
  geom_bar(stat = "identity", alpha=0.8) +
  geom_text(position = position_stack(vjust = 0.5), size = 3) +
  scale_fill_manual(values = zone_colors) +
  scale_y_continuous(labels = percent_format()) +
  theme_classic() +
  labs(y = "Percentage of Peaks", x = "Species/Condition", fill = "Zone") +
  ggtitle("Peaks Distribution Across Zones Among Species")
print(zone_x_sp_plot)
```

```{r message=FALSE, warning=FALSE, include=FALSE}
save_ggplot(zone_x_sp_plot, "zones_sp_comparative")
```

# Session info

```{r}
sessionInfo()
```