---
title: "Chromatin Dynamics with kmeans Clustering"
subtitle: "Gene regulatory dynamics during embryogenesis in Cloeon dipterum"
author: "Joan Pallares & Maria Rossello"
date: "Last run: `r format(Sys.time(), '%d/%m/%Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: flatly
    highlight: tango
---

```{r opts_chunk setup, warning=FALSE, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE,
	cache = TRUE,
	eval = FALSE)
```

**Disclaimer:**
This analysis must be run on a high-performance computing cluster or server due to its intensive computational demands, particularly due to the large scale of genomic data and the complexity of motif enrichment processes. 

## Deep tools clustering analysis

deepTools [RamÃ­rez et al., 2016](https://academic.oup.com/bioinformatics/article/32/10/1611/1742891) is a versatile software suite specifically designed for the efficient analysis of high-throughput sequencing data, such as ChIP-seq and RNA-seq. It provides a robust framework for processing large genomic datasets to visualize, compare, and understand the dynamics of gene expression and regulation across various conditions.

This section runs a script (peaks_to_bedgraph.py) to convert peaks data from a TSV file (dynamic_peaks_x_stage.tsv) to BedGraph format and saves the output in a file named dynamic_peaks.bed. Finally this data is stored as a BigWig file

```{bash}
# Convert peaks data to bedgraph format
$WD/peaks_to_bedgraph.py $WD/dynamic_peaks_x_stage.tsv $WD/dynamic_peaks.bed

# Check the output bedgraph files
ls -lh $WD/*bedgraph

## Obtain BigWig from bedGraph ##

# Create directory to store BigWig files
mkdir $WD/BIGWIG

# Loop through each bedgraph file
ls -1 $WD/*bedgraph | while read file
do
    echo "Working on this file: $file."
    
    # Ensure the bedgraph file is sorted
    file=$(basename "$file")
    sorted_file=$(echo "sorted_"$file)
    sort -k1,1 -k2,2n $WD/$file > $WD/$sorted_file

    # Convert bedgraph to BigWig format
    output_file=$(echo $file | sed 's/bedgraph/bw/g')
    bedGraphToBigWig $WD/$sorted_file $GENOME_LEN $WD/BIGWIG/$output_file
done
```

This part utilizes deepTools' computeMatrix tool to calculate scores per genome region and generate an intermediate file (compute_matrix_output.mat.gz). It takes multiple BigWig files and a BED file containing dynamic peaks (dynamic_peaks.bed) as input, specifying parameters such as bin size and number of processors.

```{bash}
# Calculate scores per genome regions and prepare an intermediate file
computeMatrix reference-point -S $FOLDER/emb4.bw $FOLDER/emb6.bw \
$FOLDER/emb8.bw $FOLDER/emb10.bw $FOLDER/emb12.bw $FOLDER/emb14.bw \
-R $WD/dynamic_peaks.bed --referencePoint center -b 1500 \
-a 1500 -bs 10 --missingDataAsZero -p 12 --samplesLabel emb4 emb6 emb8 \
emb10 emb12 emb14 -o $WD/compute_matrix_output.mat.gz
```

This segment employs deepTools' plotHeatmap tool to create a heatmap visualization from the computed matrix (compute_matrix_output.mat.gz). The heatmap is saved as a PDF file (kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.pdf), with customizations such as heatmap size, color scheme, and legend placement. Additionally, it performs k-means clustering with 6 clusters and saves the sorted regions to a BED file.

```{bash}
# Generate heatmap
plotHeatmap -m $WD/compute_matrix_output.mat.gz \
-o $WD/kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.pdf --heatmapHeight 90 \
--heatmapWidth 12 --colorList white,rebeccapurple --legendLocation best \
--whatToShow 'heatmap and colorbar' --sortUsingSample 1 --kmeans 6 \
--outFileSortedRegions $WD/kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.bed
```

This section utilizes deepTools' plotProfile tool to generate profile plots from the computed matrix (compute_matrix_output.mat.gz). The profiles are saved as a PDF file (profile_kmeans6.pdf), showing the average signal around the dynamic peaks for each cluster obtained from k-means clustering.

```{bash}
# Generate profiles
plotProfile -m $WD/compute_matrix_output.mat.gz  --kmeans 6 \
-out $WD/profile_kmeans6.pdf
```

## Homer motive enrichment

HOMER (Hypergeometric Optimization of Motif EnRichment) is a comprehensive suite specifically developed for the analysis and discovery of motif sequences in DNA, RNA, or protein datasets.

This section utilizes awk to parse the contents of the file kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.bed (output of deepTools) and separate the data into individual BED files based on the value in the 13th column. Each resulting BED file corresponds to a cluster and is saved in the cluster_bed directory.

```{bash}
mkdir $bed/kmeans/cluster_bed
awk '{if ($0 !~ /^#/) print > "'"$bed/kmeans/cluster_bed"'/"$13".bed"}' "$bed/kmeans/kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.bed"
```

This part sorts the original BED file kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.bed and separates its contents into individual BED files for each cluster, similar to the previous step. It ensures that each cluster's BED file is sorted by chromosome and genomic position.

```{bash}
sort -k 1,1 -k2,2n "$bed/kmeans/kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.bed" | awk '{if ($0 !~ /^#/) print > "'"$bed/kmeans/cluster_bed"'/"$13".bed"}'
```

This section iterates through each cluster's BED file in the cluster_bed directory. For each cluster, it performs a BED intersection using bedtools intersect to subtract regions specific to that cluster from the original BED file. The resulting BED files, named bg_cluster.bed, contain background regions not associated with the respective cluster.

```{bash}
for file in $bed/kmeans/cluster_bed/cluster_*.bed; do
    cluster=$(basename "$file" | cut -d'.' -f1)

    bedtools intersect \
        -a $bed/kmeans/kmeans6_dynamic_peaks_processed_1dot5kb_rebecca.bed \
        -b $bed/kmeans/cluster_bed/$cluster.bed  \
        -v \
        > $bed/kmeans/cluster_bed/bg_$cluster.bed
done
```

This segment loops through each cluster's BED file in the cluster_bed directory. For each cluster, it creates a directory within $homer/kmeans/ named after the cluster. Then, it executes findMotifsGenome.pl from HOMER to identify motifs in the genomic sequences represented by the cluster's BED file.

```{bash}
for file in $bed/kmeans/cluster_bed/cluster_*.bed; do
    cluster=$(basename "$file" | cut -d'.' -f1)

    mkdir $homer/kmeans/"$cluster"

    findMotifsGenome.pl \
        $bed/kmeans/cluster_bed/"$cluster".bed  \
        /data/mayfly/clodip_v3/genome_CADEPI01_2023.fasta.soft_masked \
        $homer/kmeans/"$cluster" \
        -preparsedDir $homer/clodip_v3_preparsed/ \
        -bg $bed/kmeans/cluster_bed/bg_"$cluster".bed \
        -size given \
        -mset insects \
        -p 50 \
        2> $homer/kmeans/"$cluster"/findMotifsGenome.log
done
```

